massa-execution/src/worker.rs

```diff
@@ -255,106 +253,137 @@ impl ExecutionWorker {
+            let mut cur_slot = self.last_final_slot.get_next_slot(self.cfg.thread_count)?;
+            while cur_slot <= max_css_active_slot {
+                match sce_active_blocks.first_key_value().map(|(s, _v)| *s) {
+                    // there is a CSS-active block at cur_slot
+                    Some(b_slot) if b_slot == cur_slot => {
+                        // remove the entry from sce_active_blocks (cannot panic, checked above)
+                        let (_b_slot, (b_id, block)) = sce_active_blocks.pop_first().unwrap();
+                        // execute the block as a SCE-active VM step
+                        self.push_request(ExecutionRequest::RunActiveStep(ExecutionStep {
+                            slot: cur_slot,
+                            block: Some((*b_id, block.clone())),
+                        }));
+                        // update cursor
+                        self.last_active_slot = cur_slot;
+                    }
 
-        // sort active blocks
-        self.ordered_active_blocks
-            .sort_unstable_by_key(|(_b_id, b)| b.header.content.slot);
+                    // there is no CSS-active block at cur_slot
+                    Some(_b_slot) => {
+                        // execute a miss as an SCE-active VM step
+                        self.push_request(ExecutionRequest::RunActiveStep(ExecutionStep {
+                            slot: cur_slot,
+                            block: None,
+                        }));
+                        // update cursor
+                        self.last_active_slot = cur_slot;
+                    }
 
-        // apply active blocks and misses
-        self.last_active_slot = self.last_final_slot;
-        // TODO remove clone() in iterator below
-        for (b_id, block) in self.ordered_active_blocks.clone() {
-            // process misses
-            if self.last_active_slot == self.last_final_slot {
-                self.last_active_slot =
-                    self.last_active_slot.get_next_slot(self.cfg.thread_count)?;
-            }
-            while self.last_active_slot < block.header.content.slot {
-                self.push_request(ExecutionRequest::RunActiveStep(ExecutionStep {
-                    slot: self.last_active_slot,
-                    block: None,
-                }));
-                self.last_active_slot =
-                    self.last_active_slot.get_next_slot(self.cfg.thread_count)?;
+                    // there are no more CSS-active blocks
+                    None => break,
+                }
+
+                cur_slot = cur_slot.get_next_slot(self.cfg.thread_count)?;
             }
-            self.push_request(ExecutionRequest::RunActiveStep(ExecutionStep {
-                slot: self.last_active_slot,
-                block: Some((b_id, block)),
-            }));
         }
 
-        // apply misses until now()
+        // fill with misses until current slot
         self.fill_misses_until_now()?;
```

Comment:
I think this function could be specified independently, and referred to in the spec at this point, as well. Also the current description of `fill up the remaining slots until now with misses` does not seem to match all the details that go on inside `fill_misses_until_now`.

Reply:
* I made the method description more explicit and linked it to the spec
* I made the function itself look more like the spec and added comments
* in the spec, this function now has its own "step" that can be referred to

Reply:
Great, thanks